src/ir_Mirage.cpp

```diff
@@ -67,4 +87,336 @@ bool IRrecv::decodeMirage(decode_results *results, uint16_t offset,
   // is a union data type.
   return true;
 }
+
+// Code to emulate Mirage A/C IR remote control unit.
+
+/// Class constructor
+/// @param[in] pin GPIO to be used when sending.
+/// @param[in] inverted Is the output signal to be inverted?
+/// @param[in] use_modulation Is frequency modulation to be used?
+IRMirageAc::IRMirageAc(const uint16_t pin, const bool inverted,
+                               const bool use_modulation)
+    : _irsend(pin, inverted, use_modulation) { stateReset(); }
+
+/// Reset the state of the remote to a known good state/sequence.
+void IRMirageAc::stateReset(void) {
+  // The state of the IR remote in IR code form.
+  static const uint8_t kReset[kMirageStateLength] = {
+      0x56, 0x6C, 0x00, 0x00, 0x20, 0x1A, 0x00, 0x00,
+      0x0C, 0x00, 0x0C, 0x00, 0x00, 0x00, 0x42};
+  setRaw(kReset);
+}
+
+/// Set up hardware to be able to send a message.
+void IRMirageAc::begin(void) { _irsend.begin(); }
+
+#if SEND_MITSUBISHI_AC
+/// Send the current internal state as an IR message.
+/// @param[in] repeat Nr. of times the message will be repeated.
+void IRMirageAc::send(const uint16_t repeat) {
+  _irsend.sendMirage(getRaw(), kMirageStateLength, repeat);
+}
+#endif  // SEND_MITSUBISHI_AC
+
+/// Get a PTR to the internal state/code for this protocol.
+/// @return PTR to a code for this protocol based on the current internal state.
+uint8_t *IRMirageAc::getRaw(void) {
+  checksum();
+  return _.raw;
+}
+
+/// Set the internal state from a valid code for this protocol.
+/// @param[in] data A valid code for this protocol.
+void IRMirageAc::setRaw(const uint8_t *data) {
+  std::memcpy(_.raw, data, kMirageStateLength);
+}
+
+/// Calculate and set the checksum values for the internal state.
+void IRMirageAc::checksum(void) { _.Sum = calculateChecksum(_.raw); }
+
+/// Verify the checksum is valid for a given state.
+/// @param[in] data The array to verify the checksum of.
+/// @return true, if the state has a valid checksum. Otherwise, false.
+bool IRMirageAc::validChecksum(const uint8_t *data) {
+  return calculateChecksum(data) == data[kMirageStateLength - 1];
+}
+
+/// Calculate the checksum for a given state.
+/// @param[in] data The value to calc the checksum of.
+/// @return The calculated checksum value.
+uint8_t IRMirageAc::calculateChecksum(const uint8_t *data) {
+  return sumNibbles(data, kMirageStateLength - 1);
+}
+
+/// Set the requested power state of the A/C to on.
+void IRMirageAc::on(void) { setPower(true); }
+
+/// Set the requested power state of the A/C to off.
+void IRMirageAc::off(void) { setPower(false); }
+
+/// Change the power setting.
+/// @param[in] on true, the setting is on. false, the setting is off.
+void IRMirageAc::setPower(bool on) {
+  if (on)
+    _.SwingAndPower -= getPower() ? 0 : kMirageAcPowerOff;
+  else
+    _.SwingAndPower += getPower() ? kMirageAcPowerOff : 0;
+}
```

Comment:
This looks very magicy

Reply:
Totally fair call. Stupidly long explaination for a seemingly stupid way of determining/changing the power state added in https://github.com/crankyoldgit/IRremoteESP8266/pull/1615/commits/4e4779a9b556c0499210409e88eca9baada61ba8
Please see if that explains it well enough. I'm personally not convinced the analysis is _exactly_ how it works, but their data seems to back it up. This does get the "worst possible way of turning something on/off" award so far. :-(

Reply:
Still confused, but in this case that's perfectly fine - the comment makes it clear that this is confusing. ;)